src/cpp/ext/filters/census/environment_autodetect.h

```diff
@@ -0,0 +1,71 @@
+//
+// Copyright 2023 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#ifndef GRPC_SRC_CPP_EXT_FILTERS_CENSUS_ENVIRONMENT_AUTODETECT_H
+#define GRPC_SRC_CPP_EXT_FILTERS_CENSUS_ENVIRONMENT_AUTODETECT_H
+
+#include <grpc/support/port_platform.h>
+
+#include <map>
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "absl/base/thread_annotations.h"
+#include "absl/functional/any_invocable.h"
+
+#include "src/core/lib/gprpp/sync.h"
+
+namespace grpc {
+namespace internal {
+
+class EnvironmentAutoDetect {
```

Comment:
Does this code actually belong in the census filter directory?  It seems to me that it's really a more general GCP-specific autodetection thing, not really anything specific to the census filter.  Maybe this should go in its own library somewhere?

Note that we also have some other existing code to detect GCP platform here, which we use in ALTS and in GoogleDefaultCreds:

https://github.com/grpc/grpc/blob/0ecc18ef0f51eb38880835db34f38f99980367e4/src/core/lib/security/credentials/alts/check_gcp_environment.h#L55

@apolcyn may have some thoughts about the best way to structure all of this code into a single combined library.

Reply:
Moved it to src/cpp/ext/gcp/... Alex, let's chat on whether we'd like to move some of this in Core if it's reusable.
I'd like to not block this PR though.